lib/Optimizer/Transforms/Mapping.cpp

```diff
@@ -0,0 +1,624 @@
+      phyDecay[phy1.index] += decayDelta;
+    }
+  }
+  LLVM_DEBUG(logger.startLine() << '\n' << logLineComment << '\n';);
+}
+
+//===----------------------------------------------------------------------===//
+// Pass implementation
+//===----------------------------------------------------------------------===//
+
+struct Mapper : public cudaq::opt::impl::MappingPassBase<Mapper> {
+  using MappingPassBase::MappingPassBase;
+
+  void runOnOperation() override {
+
+    // Allow debug for this pass to be enabled via environment variable
+    auto debugMappingStr = StringRef(getenv("DEBUG_MAPPING"));
+    if (debugMappingStr.equals_insensitive("1") ||
+        debugMappingStr.starts_with_insensitive("y")) {
+      const char *debugTypes[] = {DEBUG_TYPE, nullptr};
+      llvm::DebugFlag = true;
+      llvm::setCurrentDebugTypes(debugTypes, 1);
+    }
+
+    auto func = getOperation();
+    auto &blocks = func.getBlocks();
+
+    // Current limitations:
+    //  * Can only map a entry-point kernel
+    //  * The kernel can only have one block
+
+    // FIXME: Add the ability to handle multiple blocks.
+    if (blocks.size() > 1) {
+      func.emitError("The mapper cannot handle multiple blocks");
+      signalPassFailure();
+    }
+
+    // Get device
+    StringRef deviceDef = device;
+    StringRef name = deviceDef.take_front(deviceDef.find_first_of('('));
+    std::size_t deviceDim[2] = {0, 0};
+
+    // Trim the dimensions off of `deviceDef` if dimensions were provided in the
+    // string
+    if (name.size() < deviceDef.size())
+      deviceDef = deviceDef.drop_front(name.size());
+
+    StringRef deviceFilename;
+    if (name.equals_insensitive("file")) {
+      if (deviceDef.consume_front("(")) {
+        deviceDef = deviceDef.ltrim();
+        if (deviceDef.consume_back(")")) {
+          deviceFilename = deviceDef;
+          // Make sure the file exists before continuing
+          if (!llvm::sys::fs::exists(deviceFilename)) {
+            func.emitError("Path " + deviceFilename + " does not exist");
+            signalPassFailure();
+          }
+        } else {
+          func.emitError("Missing closing ')' in device option");
+          signalPassFailure();
+        }
+      } else {
+        func.emitError("Filename must be provided in device option like "
+                       "file(/full/path/to/device_file.txt): " +
+                       device.getValue());
+        signalPassFailure();
+      }
+    } else {
+      if (deviceDef.consume_front("(")) {
+        deviceDef = deviceDef.ltrim();
+        deviceDef.consumeInteger(/*Radix=*/10, deviceDim[0]);
+        deviceDef = deviceDef.ltrim();
+        if (deviceDef.consume_front(","))
+          deviceDef.consumeInteger(/*Radix=*/10, deviceDim[1]);
+        deviceDef = deviceDef.ltrim();
+        if (!deviceDef.consume_front(")")) {
+          func.emitError("Missing closing ')' in device option");
+          signalPassFailure();
+        }
+      }
+    }
```

Comment:
All this device string parsing and such should also be done when the pass is added to the pass manager and not in the pass itself (where it will be done repetitively).

Reply:
Updated in cd4eda08.